brave/src/main/java/brave/internal/recorder/MutableSpan.java

```diff
@@ -47,6 +49,22 @@ public void kind(Kind kind) {
     this.kind = kind;
   }
 
+  /** @see brave.Span#remoteServiceName(String) */
+  public void remoteServiceName(String remoteServiceName) {
+    if (remoteServiceName == null) throw new NullPointerException("remoteServiceName == null");
+    if (remoteEndpoint == null) remoteEndpoint = new MutableEndpoint();
+    remoteEndpoint.serviceName(remoteServiceName);
+  }
+
+  /** @see brave.Span#parseRemoteIpAndPort(String, int) */
+  public boolean parseRemoteIpAndPort(@Nullable String remoteIp, int remotePort) {
+    if (remoteIp == null) return false;
```

Comment:
I am curious about this interface, here you return `true/false` whereas in `removeServiceName` you throw an exception.

Reply:
if you look at the javadoc on Span.parseRemoteIpAndPort you'll notice that it is common to have to look in multiple places for ip information. Ex. here

```java
    InetAddress address = target.getAddress();
    if (address != null) {
      if (span.parseRemoteIpAndPort(address.getHostAddress(), target.getPort())) return;
    }
    span.parseRemoteIpAndPort(target.getHostName(), target.getPort());
```

This kindof code looks even worse if you are constantly checking for null. It is special cased similar to how code was done in zipkin2.Endpoint due to the crazy amount of null checking you'd need otherwise :P

remoteServiceName is like all the other fields on span.. we shouldn't set them to null. the only odd one is the remote socket info.

Reply:
for example, consider address.getHostAddress() can also return null and target.getPort() could conceivably return -1. When looking at all the call sites, it made this code least atrocious to be a validating call, notably as we don't depend on IP info anyway strictly. Also, in the case of no-op we "trick" by returning true always. This means a bit better performance when we are recording nothing. This is why there are tons of notes on the javadoc, but anyway the call sites are significantly less ugly this way imho.